### PR TITLE
Added Login Requirement for Users and Cleaned Up Code

### PR DESCRIPTION
**Pull request by javidangarcia**
- Added login requirement for users to use the app.
- Fixed comments bug where comments weren't sorted properly by date created.
- Tweaked some code based on feedback.

https://github.com/javidangarcia/stock-zone/assets/89328228/1cabaf83-5df7-4843-be11-f625ac882ea4

### CODE REVIEW

File: backend/routes/comments.js [hunk 1]
```diff
@@ -45,7 +45,9 @@ router.post("/comments", async (req, res) => {
             StockId: stock.id
         };
 
-        const comment = await Comment.create(commentData);
+        const newComment = await Comment.create(commentData);
+
+        const comment = await Comment.findOne({ where: { id: newComment.id }, include: [{ model: User }] });
```
Comment (vandanasub) [inline backend/routes/comments.js]: interesting, what exactly was the issue with `comment` before?

**Review by vandanasub — COMMENTED** [review]
great work!

File: backend/ranking.js [hunk 2]
```diff
@@ -35,23 +33,20 @@ export async function RankingOptimized(user) {
...
-
-        const likesMap = {};
-
-        likes.forEach((like) => {
-            likesMap[like.Stock.ticker] = true;
-        });
+        const followsMap = follows.reduce((accum, current) => { 
+            accum[current.Stock.ticker] = true; 
+            return accum; 
+        }, {});
```
Comment (BriptimusPrimus) [inline backend/ranking.js:39]: Looks good! Array.reduce takes us to the next level of functional programming.

File: backend/ranking.js [hunk 3]
```diff
@@ -4,13 +4,11 @@ import { Like } from "./models/like.js";
...
 import { compareStocksByPoints } from "./utils.js";
 
-
 const FOLLOW_POINTS = 25;
 const LIKE_POINTS = 20;
 const DISLIKE_POINTS = -20;
 
-
-export async function RankingOptimized(user) {
+export async function RankingV2(user) {
```
Comment (BriptimusPrimus) [inline backend/ranking.js:11]: 👍

File: backend/routes/stocks.js [hunk 4]
```diff
@@ -56,37 +57,23 @@ router.get("/stocks/:ticker", async (req, res) => {
...
+        const UserId = user.id;
 
-        if (user != null) {
-            const StockId = stock.id;
-            const UserId = user.id;
+        const [following, liking, disliking] = await Promise.all([
+            Follow.findOne({ where: { UserId, StockId } }),
+            Like.findOne({ where: { UserId, StockId } }),
+            Dislike.findOne({ where: { UserId, StockId } })
+        ]);
```
Comment (BriptimusPrimus) [inline backend/routes/stocks.js:67]: Good job. This performs a lot better than the previous version.

File: frontend/src/components/App/App.jsx [hunk 5]
```diff
@@ -24,31 +24,27 @@ export default function App() {
         localStorage.setItem("user", JSON.stringify(user));
     }, [user]);
 
+    const requireAuth = (element) => {
+        return user ? element : <LoginForm />;
```
Comment (BriptimusPrimus) [inline frontend/src/components/App/App.jsx:28]: It is better to stick to check for a boolean explicitly: `return user != null ? element : <LoginForm />;`
`user` is an object, not a boolean. By doing `user ?` you're relying in JS coercion and expecting an object to be "thruthy" and a null|undefined to be "falsy". 
That is a little hacky, and the reader has to make more effort reading it. An explicit boolean check is more clear, and recommended.

File: backend/routes/stocks.js [hunk 6]
```diff
@@ -56,37 +57,23 @@ router.get("/stocks/:ticker", async (req, res) => {
             return;
         }
 
-        const user = req.session.user;
+        const StockId = stock.id;
+        const UserId = user.id;
 
-        if (user != null) {
```
Comment (BriptimusPrimus) [inline backend/routes/stocks.js:61]: Why is this check not necessary?

File: frontend/src/components/App/App.jsx [hunk 7]
```diff
@@ -24,31 +24,27 @@ export default function App() {
...
-                        <Route
-                            path="/signup"
-                            element={<SignUpForm />}
-                        ></Route>
-                        <Route path="/profile" element={<Profile />}></Route>
-                        <Route path="/profile/:username" element={<Profile />}></Route>
+                        <Route path="/login" element={<LoginForm />} />
+                        <Route path="/signup" element={<SignUpForm />} />
+
+                        <Route path="/" element={requireAuth(<Home />)} />
```
Comment (BriptimusPrimus) [inline frontend/src/components/App/App.jsx:41]: Nice trick! 👍

File: frontend/src/components/Comments/Comments.jsx [hunk 8]
```diff
@@ -73,9 +74,9 @@ export default function Comments({ ticker }) {
         <div className="comments">
             <h2>Comments</h2>
             <div className="past-comments">
-                {comments?.map((comment) => {
+                {comments?.sort(compareCommentsByDate).map((comment) => {
                     return (
-                        <div className="comment-details" key={comment.content}>
+                        <div className="comment-details" key={`${comment.createdAt}-${comment.content}`}>
```
Comment (BriptimusPrimus) [inline frontend/src/components/Comments/Comments.jsx:79]: Nit: Is it possible to use the comment database ID here?

File: backend/routes/comments.js (hunk 1 above)
Comment (javidangarcia) [reply backend/routes/comments.js]: The newComment wouldn't include the user's information if I did include: [{model: User}] in the create function for some reason so I had to use findOne to solve that.

File: backend/routes/stocks.js (hunk 6 above)
Comment (javidangarcia) [reply backend/routes/stocks.js:61]: Users have to be logged in now to use the app and the checkSession middleware makes sure a user exists.

File: frontend/src/components/Comments/Comments.jsx (hunk 8 above)
Comment (javidangarcia) [reply frontend/src/components/Comments/Comments.jsx:79]: Good idea

**Pull request merged**